File: ts/src/principal-clone.ts
/**
 * Principal.clone helper.
 *
 * In TS we use a standalone function rather than a method-on-interface because
 * `Principal` is a plain interface with readonly reference fields — middleware
 * layers call `clonePrincipal(p)` before mutating spaceId.
 *
 * The three reference-typed fields — relatedUids, context.spaces,
 * context.ownedBotsBySpace — are copied element-wise; scalars are
 * shallow-copied by object spread.
 *
 * Nullish-safe: `clonePrincipal(undefined)` returns undefined.
 */

import type { Principal, PrincipalContext } from './verifier.js'

export function clonePrincipal(p: Principal): Principal
export function clonePrincipal(p: undefined): undefined
export function clonePrincipal(p: Principal | undefined): Principal | undefined
export function clonePrincipal(
  p: Principal | undefined,
): Principal | undefined {
  if (p === undefined) return undefined
  const dst: Principal = { ...p, context: cloneContext(p.context) }
  if (p.relatedUids !== undefined) {
    dst.relatedUids = [...p.relatedUids]
  }
  return dst
}

function cloneContext(c: PrincipalContext): PrincipalContext {
  const dst: PrincipalContext = { kind: c.kind }
  if (c.spaces !== undefined) {
    dst.spaces = [...c.spaces]
  }
  if (c.ownedBotsBySpace !== undefined) {
    const map: Record<string, readonly string[]> = {}
    for (const k of Object.keys(c.ownedBotsBySpace)) {
      const arr = c.ownedBotsBySpace[k]
      map[k] = arr === undefined ? [] : [...arr]
    }
    dst.ownedBotsBySpace = map
  }
  return dst
}

File: go/principal_clone.go
package octoauth

// Clone returns a deep copy of p safe for the caller to mutate without
// affecting other references (notably any cached *Principal that verifiers
// share across concurrent calls).
//
// The three reference-typed fields — RelatedUIDs, Context.Spaces, and
// Context.OwnedBotsBySpace — are copied element-wise; all other fields are
// value types and are shallow-copied by struct assignment.
//
// Clone is nil-safe: (*Principal)(nil).Clone() returns nil.
func (p *Principal) Clone() *Principal {
	if p == nil {
		return nil
	}
	dst := *p
	if p.RelatedUIDs != nil {
		dst.RelatedUIDs = append([]string(nil), p.RelatedUIDs...)
	}
	dst.Context = p.Context.clone()
	return &dst
}

// clone returns a deep copy of the PrincipalContext. Called by
// [Principal.Clone]; kept unexported since the middleware layer only ever
// needs to clone through Principal.
func (c PrincipalContext) clone() PrincipalContext {
	dst := c
	if c.Spaces != nil {
		dst.Spaces = append([]string(nil), c.Spaces...)
	}
	if c.OwnedBotsBySpace != nil {
		dst.OwnedBotsBySpace = make(map[string][]string, len(c.OwnedBotsBySpace))
		for k, v := range c.OwnedBotsBySpace {
			if v == nil {
				dst.OwnedBotsBySpace[k] = nil
				continue
			}
			dst.OwnedBotsBySpace[k] = append([]string(nil), v...)
		}
	}
	return dst
}
